File: simulator/worker.py
from messenger import Messenger


class Sensors:

    def __init__(self, maze, id):
        self.maze = maze
        self.id = id


    def get_sensors_vector(self):
        return self.maze.get_sensors_vector(self.id)


    def get_situation(self):
        (current_path, left_path, right_path, front_path) = self.get_sensors_vector()

        if (left_path == 0 and current_path == 1 and
            right_path == 0 and front_path == 1):
            return "normal"
        else:
            return "intersection"


    def get_type_intersection(self):
        (current_path, left_path, right_path, front_path) = self.get_sensors_vector()

        if (left_path and front_path and right_path):
            return('A')
        elif (left_path and front_path and not right_path):
            return('B')
        elif (not left_path and front_path and right_path):
            return('C')
        elif (left_path and not front_path and right_path):
            return('D')
        elif (left_path and not front_path and not right_path):
            return('E')
        elif (not left_path and not front_path and right_path):
            return('F')
        elif (not left_path and not front_path and not right_path):
            return('G')
        return('')


class Actuators:

    def __init__(self, maze, id):
        self.maze = maze
        self.id = id


    def go_left(self):
        self.maze.turn_left(self.id)
        self.go_straight()


    def go_straight(self):
        self.maze.go_straight(self.id)


    def go_right(self):
        self.maze.turn_right(self.id)
        self.go_straight()


    def go_back(self):
        self.maze.turn_back(self.id)
        self.go_straight()


    def stop(self):
        return None


class Worker:

    def __init__(self, maze, id):
        self.stop_signal = False
        self.id = id
        self.sensors = Sensors(maze, id)
        self.actuators = Actuators(maze, id)
        self.messenger = None
        self.distance = 0
        

    def get_id(self):
        return self.id


    def connect(self, messenger, master_id):
        self.master_id = master_id
        self.messenger = messenger


    def receive_msg(self, sender_id, type_message, content_message):
        if type_message == "instruction":
            self.stop_signal = True


    def send_msg(self, receiver_id, type_message, content_message):
        self.messenger.send(receiver_id, type_message,
            content_message)


    def run(self):
        if not self.stop_signal:
            situation = self.sensors.get_situation()

            if situation == "normal":
                self.actuators.go_straight()

            elif situation == "not-valid":
                self.actuators.stop()
                self.stop_signal = True

            elif situation == "intersection":
                self.actuators.stop()
                type_intersection = self.sensors.get_type_intersection()
                new_direction = self.direction(type_intersection)
                if new_direction == "left":
                    self.actuators.go_left()
                elif new_direction == "straight":
                    self.actuators.go_straight()
                elif new_direction == "right":
                    self.actuators.go_right()
                elif new_direction == "uturn":
                    self.actuators.go_back()
                else:
                    self.actuators.stop()
                    self.stop_signal = True

                # Send message to master (map information)
                content_message = {"traveled_distance": self.distance,
                    "taken_direction": new_direction}
                self.send_msg(self.master_id, "map_info", content_message)
                self.distance = 0


    def direction(self, type_intersection):
        if type_intersection == 'A':
            return "left"
        elif type_intersection == 'B':
            return "left"
        elif type_intersection == 'C':
            return "straight"
        elif type_intersection == 'D':
            return "left"
        elif type_intersection == 'E':
            return "left"
        elif type_intersection == 'F':
            return "right"
        elif type_intersection == 'G':
            return "uturn"
        return "stop"

File: worker/src/Worker.cpp


#include "../include/Worker.h"


Worker::Worker(const int& id) {
    this->id = id;
    this->distance = 0;
    this->run = true;
    this->sensors = new Sensors();
    this->actuators = new Actuators();
    this->time_step = 2;  // In milliseconds
}


void Worker::Run() {
    while (this->run) {
        this->distance++;
        std::string situation = this->sensors->GetSituation()

        if (situation == "normal") 
            this->actuators->GoStraight();

        else if (situation == "not-valid") {
            this->Send(0, "info-map-distance", this->distance);
            this->distance = 0;
            this->actuators->Stop();
            this->run = false;
        }

        else if (situation == "intersection") {
            this->Send(0, "info-map-distance", this->distance);
            this->distance = 0;

            // Read sensors to detemine type of encountered intersection 
            this->actuators->Stop();
            int* sensors_vector = this->sensors->GetVector();
            this->actuators->OneInch();
            int* sensors_vector_one_inch = this->sensors->GetVector();

            // Determine type of intersection
            char type_intersection = this->sensors->GetTypeIntersection(
                sensors_vector, sensors_vector_one_inch);

            // Determine direction to take
            std::string new_direction = this->ComputeDirection(type_intersection);

            // Take direction
            if (new_direction == "left") 
                this->actuators->GoLeft();
            else if (new_direction == "straight")
                this->actuators->GoStraight();
            else if (new_direction == "right")
                this->actuators->GoRight();
            else if (new_direction == "uturn")
                this->actuators->GoBack();
            else {
                this->actuators->Stop();
                this->run = false;
            }
        }

        usleep(this->time_step);

        // Compute traveled distance
        float applied_velocity = this->actuators->GetVelocity();
        float traveled_distance = this->ComputeDistance(applied_velocity);
        this->distance += applied_distance;
    }
}


int Worker::GetId() {
    return this->id;
}


std::string Worker::ComputeDirection(char type_intersection) {
    if (type_intersection == 'A')
        return "left"
    else if (type_intersection == 'B')
        return "left"
    else if (type_intersection == 'C')
        return "straight"
    else if (type_intersection == 'D')
        return "left"
    else if (type_intersection == 'E')
        return "left"
    else if (type_intersection == 'F')
        return "right"
    else if (type_intersection == 'G')
        return "uturn"
    return "stop"
} 


float Worker::ComputeDistance(float velocity) {
    float time_step_in_seconds = ((float) this->time_step) / 1000.;
    return velocity * time_step_in_seconds;  // In cm
}


void Worker::Send(int receiver_id, std::string type_message,
        float content_message) {
    std::cout << "Receiver ID: " << receiver_id << std::endl;
    std::cout << "Type Message: " << type_message << std::endl;
    std::cout << content_message << std::endl;
}









